config_parser.cc

```diff
@@ -194,19 +195,21 @@ bool NginxConfigParser::Parse(std::istream* config_file, NginxConfig* config) {
         // Error.
         break;
       }
+      bracket_match++;
       NginxConfig* const new_config = new NginxConfig;
       config_stack.top()->statements_.back().get()->child_block_.reset(
           new_config);
       config_stack.push(new_config);
     } else if (token_type == TOKEN_TYPE_END_BLOCK) {
-      if (last_token_type != TOKEN_TYPE_STATEMENT_END) {
+      if (last_token_type != TOKEN_TYPE_STATEMENT_END && last_token_type != TOKEN_TYPE_END_BLOCK) {
         // Error.
         break;
       }
+      bracket_match--;
       config_stack.pop();
     } else if (token_type == TOKEN_TYPE_EOF) {
       if (last_token_type != TOKEN_TYPE_STATEMENT_END &&
-          last_token_type != TOKEN_TYPE_END_BLOCK) {
+          last_token_type != TOKEN_TYPE_END_BLOCK || bracket_match != 0) {
```

Comment:
I think it should be && instead of ||. Right now I think it will pass if the last token before the EOF is a semicolon (token_type_statement_end) or a closed brace (token_type_end_block). Try a test where you have mismatched braces, but the last character is a bracket (e.g. 'test { something { foo bar;}')